### ISSUE TITLE
RHEL bootstrap repos not getting created for Uyuni

### ISSUE THREAD

**Issue by EthanB11** [uyuni-project/uyuni#2096]
Uyuni 2020.3

Trying to bootstrap RHEL7 from the web results in this error:
file_|-/etc/salt/minion_id_|-/etc/salt/minion_id_|-managed(retcode=2): One or more requisite failed: bootstrap.salt-minion-package service_|-salt-minion_|-salt-minion_|-running(retcode=2): One or more requisite failed: bootstrap./etc/salt/minion_id, bootstrap.salt-minion-package, bootstrap./etc/salt/pki/minion/minion.pub, bootstrap./etc/salt/pki/minion/minion.pem, bootstrap./etc/salt/minion.d/susemanager.conf file_|-/etc/salt/minion.d/susemanager.conf_|-/etc/salt/minion.d/susemanager.conf_|-managed(retcode=2): One or more requisite failed: bootstrap.salt-minion-package file_|-/etc/salt/pki/minion/minion.pem_|-/etc/salt/pki/minion/minion.pem_|-managed(retcode=2): One or more requisite failed: bootstrap.salt-minion-package pkg_|-salt-minion-package_|-salt-minion_|-latest(retcode=2): No information found for 'salt-minion'. file_|-/etc/salt/pki/minion/minion.pub_|-/etc/salt/pki/minion/minion.pub_|-managed(retcode=2): One or more requisite failed: bootstrap.salt-minion-package

On the server Fedora and RHEL bootstrap repos are being created under unknown.

````
foohost:~ # ls /srv/www/htdocs/pub/repositories/
centos/    empty/     empty-deb/ opensuse/  sle/       ubuntu/    unknown/
foohost:~ # ls /srv/www/htdocs/pub/repositories/unknown/
29  7
````

**Comment by EthanB11** [comment]
Here is the output of running the bootstrap script locally. It's looking for res-7 when it does not exist.

````
[root@rhel7 ~]# sh bootstrap-rhel7-dev-test.sh
Uyuni Server Client bootstrap script v4.1


MINOR MANUAL EDITING OF THIS FILE MAY BE REQUIRED!

If this bootstrap script was created during the initial installation
of a Uyuni Server, the ACTIVATION_KEYS, and ORG_GPG_KEY values will
probably *not* be set (see below). If this is the case, please do the
following:
  - copy this file to a name specific to its use.
    (e.g., to bootstrap-SOME_NAME.sh - like bootstrap-web-servers.sh.)
  - on the website create an activation key or keys for the system(s) to
    be registered.
  - edit the values of the VARIABLES below (in this script) as
    appropriate:
    - ACTIVATION_KEYS needs to reflect the activation key(s) value(s)
      from the website. XKEY or XKEY,YKEY
      Please note that if you are using this script to boostrap minions,
      only the FIRST activation key will be used. Multiple activation keys
      are not supported with salt
    - ORG_GPG_KEY needs to be set to the name(s) of the corporate public
      GPG key filename(s) (residing in /srv/www/htdocs/pub) if appropriate. XKEY or XKEY,YKEY

Verify that the script variable settings are correct:
    - CLIENT_OVERRIDES should be only set differently if a customized
      client-config-overrides-VER.txt file was created with a different
      name.
    - ensure the value of HOSTNAME is correct.
    - ensure the value of ORG_CA_CERT is correct.

Enable this script: comment (with #'s) this block (or, at least just
the exit below, if present)


UPDATING RHN_REGISTER/UP2DATE CONFIGURATION FILES
-------------------------------------------------
* downloading necessary files
  client_config_update.py...
2020-04-03 08:59:14 URL:https://foohost.foo.domain.com/pub/bootstrap/client_config_update.py [6456/6456] -> "client_config_update.py" [1]
FINISHED --2020-04-03 08:59:14--
Total wall clock time: 0.01s
Downloaded: 1 files, 6.3K in 0s (353 MB/s)
  client-config-overrides.txt...
2020-04-03 08:59:14 URL:https://foohost.foo.domain.com/pub/bootstrap/client-config-overrides.txt [627/627] -> "client-config-overrides.txt" [1]
FINISHED --2020-04-03 08:59:14--
Total wall clock time: 0.01s
Downloaded: 1 files, 627 in 0s (78.9 MB/s)

PREPARE GPG KEYS AND CORPORATE PUBLIC CA CERT
-------------------------------------------------

* importing organizational GPG keys
2020-04-03 08:59:15 URL:https://foohost.foo.domain.com/pub/uyuni-gpg-pubkey-0d20833e.key [1126/1126] -> "uyuni-gpg-pubkey-0d20833e.key" [1]
FINISHED --2020-04-03 08:59:15--
Total wall clock time: 0.01s
Downloaded: 1 files, 1.1K in 0s (336 MB/s)

* attempting to install corporate public CA cert
2020-04-03 08:59:15 URL:https://foohost.foo.domain.com/pub/rhn-org-trusted-ssl-cert-1.0-1.noarch.rpm [13572/13572] -> "rhn-org-trusted-ssl-cert-1.0-1.noarch.rpm" [1]
FINISHED --2020-04-03 08:59:15--
Total wall clock time: 0.01s
Downloaded: 1 files, 13K in 0s (271 MB/s)
Preparing...                          ################################# [100%]
Updating / installing...
   1:rhn-org-trusted-ssl-cert-1.0-1   ################################# [100%]

CLEANING UP OLD SUSE MANAGER REPOSITORIES
-------------------------------------------------

CHECKING THE REGISTRATION STACK
-------------------------------------------------
* check for necessary packages being installed...
* client codebase is res-7
package salt is not installed
package salt-minion is not installed
https://foohost.foo.domain.com/pub/repositories/res/7/bootstrap/repodata/repomd.xml:
2020-04-03 08:59:16 ERROR 404: 404.
Bootstrap repo 'https://foohost.foo.domain.com/pub/repositories/res/7/bootstrap' does not exist.
* going to install missing packages...
Loaded plugins: langpacks, product-id, search-disabled-repos, subscription-manager
This system is not registered with an entitlement server. You can use subscription-manager to register.
There are no enabled repos.
 Run "yum repolist all" to see the repos you have.
 To enable Red Hat Subscription Management repositories:
     subscription-manager repos --enable <repo>
 To enable custom repositories:
     yum-config-manager --enable <repo>
package salt is not installed
ERROR: Failed to install all missing packages.
````

**Comment by juliogonzalez** [comment]
Can you please try to recreate the bootstrap repository, this time with `-d`,  and paste the log?

Also use `-f`.

```
mgr-create-bootstrap-repo -d -f
```
And the rest of the usual parameters.

**Comment by EthanB11** [comment]
I can't get it working at all.

I set up the main channel for RHEL as shown here. https://www.uyuni-project.org/uyuni-docs/uyuni/client-configuration/clients-rh.html

**Available Bootstrap Repos**
````
foohost:~ # mgr-create-bootstrap-repo -l
1. SLE-15-SP1-x86_64
2. centos-7-x86_64
3. centos-8-x86_64
4. openSUSE-Leap-15.1-x86_64
5. ubuntu-18.04-amd64-uyuni
````

**Create bootstrap repo with parent channel**
````
foohost:~ # mgr-create-bootstrap-repo  -d -f --with-parent-channel rhel7-x86_64-server -c RES7-x86_64
'rhel7-x86_64-server' not found in existing parent channel options '[]'
````

````
foohost:~ # tail  /var/log/rhn/mgr-create-bootstrap-repo.log
2020/04/06 14:06:37 -05:00 33543 0.0.0.0: rhnSQL/driver_postgresql._execute_wrapper('Executing SQL: "\nSELECT sp.product_id\n  FROM suseProducts sp\n  JOIN rhnChannelFamily cf ON sp.channel_family_id = cf.id\n WHERE cf.label = \'SLE-M-T\';\n" with bind params: {}',)
2020/04/06 14:06:37 -05:00 33543 0.0.0.0: rhnSQL/driver_postgresql._execute_wrapper('Executing SQL: "\nSELECT label\n FROM rhnChannel\nWHERE id IN (\n  SELECT DISTINCT c.parent_channel\n    FROM rhnChannel c\n    JOIN suseProductChannel pc ON pc.channel_id = c.id\n    JOIN suseProducts p ON pc.product_id = p.id\n   WHERE p.product_id IN ( 1251 )\n)\n" with bind params: {}',)
2020/04/06 14:06:37 -05:00 33543 0.0.0.0: rhnSQL/driver_postgresql._execute_wrapper('Executing SQL: "\nSELECT label\n FROM rhnChannel\nWHERE id IN (\n  SELECT DISTINCT c.parent_channel\n    FROM rhnChannel c\n    JOIN suseProductChannel pc ON pc.channel_id = c.id\n    JOIN suseProducts p ON pc.product_id = p.id\n   WHERE p.product_id IN ( 1251 )\n)\n" with bind params: {}',)
2020/04/06 14:06:37 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(382) - 'rhel7-x86_64-server' not found in existing parent channel options '[]'
2020/04/06 14:08:03 -05:00: /usr/sbin/mgr-create-bootstrap-repo.cli(256) - ['/usr/sbin/mgr-create-bootstrap-repo', '-d', '-f', '--with-parent-channel', 'rhel7-x86_64-server', '-c', 'RES7-x86_64']
2020/04/06 14:08:03 -05:00: /usr/sbin/mgr-create-bootstrap-repo.cli(262) - Is a Uyuni installation
2020/04/06 14:08:03 -05:00 34125 0.0.0.0: rhnSQL/driver_postgresql._execute_wrapper('Executing SQL: "\nSELECT sp.product_id\n  FROM suseProducts sp\n  JOIN rhnChannelFamily cf ON sp.channel_family_id = cf.id\n WHERE cf.label = \'SLE-M-T\';\n" with bind params: {}',)
2020/04/06 14:08:03 -05:00 34125 0.0.0.0: rhnSQL/driver_postgresql._execute_wrapper('Executing SQL: "\nSELECT label\n FROM rhnChannel\nWHERE id IN (\n  SELECT DISTINCT c.parent_channel\n    FROM rhnChannel c\n    JOIN suseProductChannel pc ON pc.channel_id = c.id\n    JOIN suseProducts p ON pc.product_id = p.id\n   WHERE p.product_id IN ( 1251 )\n)\n" with bind params: {}',)
2020/04/06 14:08:03 -05:00 34125 0.0.0.0: rhnSQL/driver_postgresql._execute_wrapper('Executing SQL: "\nSELECT label\n FROM rhnChannel\nWHERE id IN (\n  SELECT DISTINCT c.parent_channel\n    FROM rhnChannel c\n    JOIN suseProductChannel pc ON pc.channel_id = c.id\n    JOIN suseProducts p ON pc.product_id = p.id\n   WHERE p.product_id IN ( 1251 )\n)\n" with bind params: {}',)
2020/04/06 14:08:03 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(382) - 'rhel7-x86_64-server' not found in existing parent channel options '[]'
````

**Create bootstrap repo without parent channel**
````
foohost:~ # mgr-create-bootstrap-repo -c RES7-x86_64
WARNING: no parent channel found. Execute without using custom channels

Creating bootstrap repo for latest Service Pack of RES7-x86_64
Directory walk started
Directory walk done - 0 packages
Temporary output repo path: /srv/www/htdocs/pub/repositories/res/7/bootstrap.tmp/.repodata/
Preparing sqlite DBs
Pool started (with 5 workers)
Pool finished
ERROR: package 'salt' not found
ERROR: package 'salt-minion' not found
ERROR: package 'python2-salt' not found
ERROR: package 'python-futures' not found
ERROR: package 'python-jinja2' not found
ERROR: package 'python-msgpack-python' not found
ERROR: package 'python-psutil' not found
ERROR: package 'python-pycrypto' not found
ERROR: package 'python-requests' not found
ERROR: package 'python-setuptools' not found
ERROR: package 'python-tornado' not found
ERROR: package 'python-zmq' not found
ERROR: package 'zeromq' not found
ERROR: package 'python-backports-ssl_match_hostname' not found
ERROR: package 'python-backports' not found
ERROR: package 'python-certifi' not found
ERROR: package 'python-simplejson' not found
ERROR: package 'PyYAML' not found
ERROR: package 'python-markupsafe' not found
ERROR: package 'python-urllib3' not found
ERROR: package 'libyaml' not found
ERROR: package 'python-chardet' not found
ERROR: package 'python-six' not found
ERROR: package 'python-babel' not found
ERROR: package 'yum-rhn-plugin' not found
ERROR: package 'yum' not found
ERROR: package 'python2-rhnlib' not found
ERROR: package 'openssl' not found
ERROR: package 'openssl-libs' not found
ERROR: package 'python-ipaddress' not found
ERROR: package 'rpm-python' not found
ERROR: package 'spacewalk-check' not found
ERROR: package 'python2-spacewalk-check' not found
ERROR: package 'spacewalk-client-setup' not found
ERROR: package 'python2-spacewalk-client-setup' not found
ERROR: package 'spacewalk-client-tools' not found
ERROR: package 'python2-spacewalk-client-tools' not found
ERROR: package 'python2-uyuni-common-libs' not found
ERROR: none of 'mgr-daemon', 'spacewalksd' found
ERROR: package 'suseRegisterInfo' not found
ERROR: package 'python2-suseRegisterInfo' not found
ERROR: package 'python2-hwdata' not found
ERROR: package 'dmidecode' not found

Suggestions:
mgr-create-bootstrap-repo uses the locally synchronized versions of files
from the Tools repository, and uses the locally synchronized pool channel
for dependency resolution.
Both should be fully synced before running the mgr-create-bootstrap-repo script.
````

````
foohost:~ # tail -n 55  /var/log/rhn/mgr-create-bootstrap-repo.log
2020/04/06 14:08:03 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(382) - 'rhel7-x86_64-server' not found in existing parent channel options '[]'
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.cli(256) - ['/usr/sbin/mgr-create-bootstrap-repo', '-c', 'RES7-x86_64']
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.cli(262) - Is a Uyuni installation
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(394) - WARNING: no parent channel found. Execute without using custom channels
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(414) - Creating bootstrap repo for latest Service Pack of RES7-x86_64
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'salt' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'salt-minion' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python2-salt' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-futures' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-jinja2' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-msgpack-python' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-psutil' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-pycrypto' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-requests' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-setuptools' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-tornado' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-zmq' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'zeromq' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-backports-ssl_match_hostname' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-backports' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-certifi' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-simplejson' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'PyYAML' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-markupsafe' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-urllib3' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'libyaml' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-chardet' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-six' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-babel' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'yum-rhn-plugin' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'yum' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python2-rhnlib' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'openssl' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'openssl-libs' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python-ipaddress' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'rpm-python' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'spacewalk-check' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python2-spacewalk-check' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'spacewalk-client-setup' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python2-spacewalk-client-setup' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'spacewalk-client-tools' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python2-spacewalk-client-tools' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python2-uyuni-common-libs' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: none of 'mgr-daemon', 'spacewalksd' found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'suseRegisterInfo' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python2-suseRegisterInfo' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'python2-hwdata' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(504) - ERROR: package 'dmidecode' not found
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(509) -
Suggestions:
2020/04/06 14:10:42 -05:00: /usr/sbin/mgr-create-bootstrap-repo.create_repo(511) - mgr-create-bootstrap-repo uses the locally synchronized versions of files
from the Tools repository, and uses the locally synchronized pool channel
for dependency resolution.
Both should be fully synced before running the mgr-create-bootstrap-repo script.
````

**Comment by juliogonzalez** [comment]
If Red Hat 7 is not even detected, no wonder you can't generate the repository. Even if you force the parent, the bootstrap repo generator will not know what to do.

What's the label of your parent channel? I guess `rhel7-x86_64-server` as per doc, right?

I think I see the problem at https://github.com/uyuni-project/uyuni/blob/master/susemanager/src/mgr_bootstrap_data.py#L1063
```
    'RHEL7-x86_64' : {
        'PDID' : [-7, 1683], 'BETAPDID' : [2065], 'PKGLIST' : RES7,
        'DEST' : '/srv/www/htdocs/pub/repositories/res/7/bootstrap/'
    },
```
It seems to me we need another dictionary like this, but when the base channel is `rhel7-x86_64-server`. Just as we have two pairs of dictionaries for Ubuntu 16.04 and Ubuntu 18.04 at https://github.com/uyuni-project/uyuni/blob/master/susemanager/src/mgr_bootstrap_data.py#L1075

@mcalmer can you confirm?

If so, I can prepare a PR and tell you how to apply the patch until the next Uyuni release is out.

**Comment by EthanB11** [comment]
That's correct `rhel7-x86_64-server`. What I would like to see is a common channel structure setup for RHEL like it is for other distros. I understand that the repo's can't be populated, but could the proper base channel and Uyuni client channel be created so give everyone a starting point? I think that would make things alot simpler. 

DOH! I should have looked at the mgr_bootstrap_data.py as I was updating that for Fedora....

**Comment by EthanB11** [comment]
That's correct `rhel7-x86_64-server`. What I would like to see is a common channel structure setup for RHEL like it is for other distros. I understand that the repo's can't be populated, but could the proper base channel and Uyuni client channel be created so give everyone a starting point? I think that would make things alot simpler. 

There are rhel-pool-bases that are able to be set up from the setup wizard. If those could be modified to be the proper base channels mgr-create-bootstrap-repo is looking for I think that might totally solve it. Then we can add our children and the correct repos from there.

DOH! I should have looked at the mgr_bootstrap_data.py as I was updating that for Fedora....

**Comment by juliogonzalez** [comment]
> What I would like to see is a common channel structure setup for RHEL like it is for other distros. I understand that the repo's can't be populated, but could the proper base channel and Uyuni client channel be created so give everyone a starting point? I think that would make things alot simpler.

That should be doable, yes. The only restriction is that right now we can't use the same labels from channels from SCC and from channels from `spacewalk-common-channels`.

For that, three things are needed:

1) Adding the new parent channels to https://github.com/uyuni-project/uyuni/blob/master/utils/spacewalk-common-channels.ini (one for each RHEL version). Basically the same channel labels you get from the wizzard with `-uyuni` suffix. Also the client tools need to be added be added. URLs for CentOS client tools can be used, so we'd just need copies of `[centos8-uyuni-client]` and `[centos8-uyuni-client]` replacing `centos8` with `rhel8` (same for each RHEL version).
2) Modifying https://github.com/uyuni-project/uyuni/blob/master/susemanager/src/mgr_bootstrap_data.py to support such new base channels.
3) Modifying the doc to explain this, only for Uyuni.

If I assist you with this when you have questions, can you try to take care of creating the PR? :-)

**Comment by EthanB11** [comment]
If that's all that is required I will add that to the Fedora repos and just put that all in one PR if that's ok with you?

**Comment by mcalmer** [comment]
> If Red Hat 7 is not even detected and offered as an option, no wonder you can't generate the repository. Even if you force the parent, the bootstrap repo generator will not know what to do.
> 
> What's the label of your parent channel? I guess `rhel7-x86_64-server` as per doc, right?
> 
> I think I see the problem at https://github.com/uyuni-project/uyuni/blob/master/susemanager/src/mgr_bootstrap_data.py#L1063
> 
> ```
>     'RHEL7-x86_64' : {
>         'PDID' : [-7, 1683], 'BETAPDID' : [2065], 'PKGLIST' : RES7,
>         'DEST' : '/srv/www/htdocs/pub/repositories/res/7/bootstrap/'
>     },
> ```
> 
> It seems to me we need another dictionary like this, but when the base channel is `rhel7-x86_64-server`. Just as we have two pairs of dictionaries for Ubuntu 16.04 and Ubuntu 18.04 at https://github.com/uyuni-project/uyuni/blob/master/susemanager/src/mgr_bootstrap_data.py#L1075 (those with PDID are for SUSE Manager, those with BASECHANNEL are for Uyuni).
> 
> @mcalmer can you confirm?

I can confirm that we should review the redhat bootstrap repos, but I am not sure how many new entries we need.

The problem is, that this was made for SCC and original only for SLES-ES (Expanded Support).
It was enhanced by Native RedHat, Centos for Uyuni, and next would be Centos with SCC data.

All this did not make it easier to understand the structure and we cannot "change" the structure as it would break things for our existing customers.

Btw: `rhel7-x86_64-server` does not exist. We do not use this channel.

For "7" we have the following situation:

`rhel-x86_64-server-7` is reserved for "Expanded Support". But we miss it in bootstrap data. Currently you cannot create a bootstrap repo for real Expanded Support 7.

Native RedHat 7 should use `rhel7-pool-x86_64` as base channel. This is what is defined with the entry you pasted above. 

CentOS7 for Uyuni uses another base channel. `centos7-x86_64`.
CentOS7 with SCC data is not yet defined/implemented.

In general: The bootstrap repo is optional. You only need it when the software which is required to register the client at Uyuni, is not yet installed. spacewalk never had this. So people somehow survived without this feature :-)

To find the software which is needed for such a repo and which channels to take into account is not easy. So every self invented channel structure cannot work out of the box.

But in the end you need to adapt the data file mgr_bootstrap_data.py. Uyuni support to specify a different name. So you can easy copy the file and enhance it with whatever structure you want.

```
$> mgr-create-bootstrap-repo --help
...
  --datamodule=DATAMODULE
                        Use an own datamodule (Default: mgr_bootstrap_data)
...
```

In the next version also via rhn.conf

```
server.susemenegr.bootstrap_repo_datamodule = mgr_bootstrap_data
```

**Comment by EthanB11** [comment]
> In general: The bootstrap repo is optional. You only need it when the software which is required to register the client at Uyuni, is not yet installed. spacewalk never had this. So people somehow survived without this feature :-)

Using a bootstrap from the webserver greatly reduces time to get a server in registered and improves the process for admins. Why wouldn't we want that? I had roughly ten steps I had to log into each server to do to get a system registered on spacewalk. As long as Uyuni has the ability to do this we should be using those features.

> Native RedHat 7 should use `rhel7-pool-x86_64` as base channel. This is what is defined with the entry you pasted above.

I'm pretty sure this is where my `unknown/7` and `unknown/29` bootstrap repos came from. Using the channels from the setup wizard for RH7 and Fedora29.........

How does the `mgr-create-bootstrap-repo -l` list get populated? I never saw anythign RH related when I used the `rhel7-pool-x86_64` base channel.

````
foohost:~ # mgr-create-bootstrap-repo -l
1. SLE-15-SP1-x86_64
2. centos-7-x86_64
3. centos-8-x86_64
4. openSUSE-Leap-15.1-x86_64
5. ubuntu-18.04-amd64-uyuni
````

**Comment by paususe** [comment]
Hey @EthanB11 

Are you still seeing this with Uyuni 2020.05?

**Comment by EthanB11** [comment]
I had edited the local files to force the creation, I just updated to
2020.05 yesterday so I can take a look next week.

**Comment by EthanB11** [comment]
I moved to 2020.06 and all the bootstrap repos are there, but still having the issue where it pulls the python3-tornado version 6, but I worked around that with the bootstrap script. I just have it manually install the correct python3-tornado first. I added Ubuntu 20.04 repos and sync'd them, no bootstrap repo was automatically created. I don't see an option to create a repo for it though.

````
uyuni:/srv/www/htdocs/pub/repositories # mgr-create-bootstrap-repo
1. RHEL7-x86_64-uyuni
2. SLE-15-SP1-x86_64
3. centos-7-x86_64
4. centos-8-x86_64
5. fedora31-x86_64
6. openSUSE-Leap-15.1-x86_64
7. ubuntu-18.04-amd64-uyuni
Enter a number of a product label:
````

**Comment by juliogonzalez** [comment]
Some good news: the problem with python3-tornado will go away with 2020.07, as salt 3000 will have it vendored :-)

**Comment by juliogonzalez** [comment]
As for Ubuntu 20.04, it will be available with 2020.07 as well.

**Comment by mcalmer** [comment]
I think this problem should be solved with 2020.11 . If there are new problems with the latest version, please open a new issue.
Thanks.
